### ISSUE TITLE
[Bug]: `destructuring-assignment` enforces destructuring of modern context

### ISSUE THREAD

**Issue by billyjanitsch** [jsx-eslint/eslint-plugin-react#3536]
### Is there an existing issue for this?

- [X] I have searched the existing issues and my issue is unique
- [X] My issue appears in the command-line and not only in the text editor

### Description Overview

https://github.com/jsx-eslint/eslint-plugin-react/pull/2797 was recently merged and released. That PR requires any value returned from `useContext` to be destructured if it's an object:

```js
function MyComponent() {
  const foo = useContext(FooContext)
  return <>{foo.val}</> // error
}
```

It seems like there may have been some confusion or miscommunication here. That PR was meant to close https://github.com/jsx-eslint/eslint-plugin-react/issues/2309, but as noted in https://github.com/jsx-eslint/eslint-plugin-react/issues/2309#issuecomment-674361485, the rule is only supposed to warn for destructuring of _legacy_ context (i.e., not `useContext`).

As I understand it, the spirit of the rule is to require destructuring of the plain objects that React creates. That includes `props`, as well as legacy `state` and `context` (which are always stored in objects). But it doesn't seem like it should include modern context, because context can store any (single) value. That value may or may not be an object, and even if it is an object, it's not necessarily a plain object that serves as a container for conceptually separate values. For example, it could be an instance of a class (`const logger = useContext(LoggerContext)`) in which case destructuring doesn't make sense (or, at least, it doesn't seem within the spirit/scope of this rule to enforce that variety of destructuring).

The new behavior also results in an inconsistency between modern state and modern context, even though these are conceptually the same in terms of being able to store any single value:

```js
function MyComponent() {
  const foo = useContext(FooContext)
  const bar = useState({val: 1})
  return (
    <>
      {foo.val} // error
      {bar.val} // no error
    </>
  )
}
```

### Expected Behavior

I would expect this rule not to warn on values returned from `useContext` (or `useState`).

### eslint-plugin-react version

v7.32.2

### eslint version

v8.34.0

### node version

v18.12.0

**Comment by ljharb** [comment]
That's a fair point, and I'd support a PR that adds/updates test cases (and makes whatever fixes are needed) to reflect this viewpoint.
